client-state-machine.go

```diff
@@ -102,6 +102,24 @@ func (state clientStateStart) Next(hr handshakeMessageReader) (HandshakeState, [
 		alpn = &ALPNExtension{Protocols: state.Opts.NextProtos}
 	}
 
+	// Raw public keys
+	var cct *ClientCertTypeExtension
+	var sct *ClientCertTypeExtension
+	if state.Config.AllowRawPublicKeys || !state.Config.ForbidCertificates {
```

Comment:
I don't see how that would arise.  If we take the branch, then one of those two conditions is true.  If the former, then `CertificateTypeRawPublicKey` is added; if the latter, then `CertificateTypeX509` is added.  So either you have at least one entry or `cct == sct == nil`, and the logic below does not add them if they are nil.

Reply:
Maybe I don't understand what Config.ForbidCertificates is supposed to do.

Reply:
`Config.ForbidCertificates = true` means that `CertificateTypeX509` is not included in the certificate type extensions, if they are sent.  I agree that's confusing.  ISTM there are two plausible refactors here, both assuming removal of `ForbidCertificates`:

1. Assume that X.509 is allowed, so `AllowRawPublicKeys` means that the cert type extension will have both values.  This is simple, but constrains the options available to the application.
2. Just have explicit type arrays in the `Config` object.  This is more flexible, but we might have to check for inconsistent config.

Reply:
I think I'm missing something, because if !AllowedRawPublicKeys && ForbidCertificates, then isn't the handshake impossible? So aren't you sending the extension for any valid config?

Reply:
True, it'll just fail later.  And obviously, in the default X.509-only case, we don't need to send the extensions at all.  So it seems like the correct logic here (without changing the config format) would be something like:

```
if state.Config.AllowRawPublicKeys {
    // create extensions with RPKs allowed
    if !state.Config.ForbidCertificates {
        // add X509 to extensions
    }
}
```

But of course that muddies the definition of ForbidCertificates. Maybe the right refactor here is `ForbidCertificates -> RequireRawPublicKeys`, and:


```
if state.Config.AllowRawPublicKeys {
    // create extensions with RPKs allowed
    if !state.Config.RequireRawPublicKeys {
        // add X509 to extensions
    }
}
```

Reply:
I think I prefer the first, but I think either would work.